state-chain/pallets/cf-elections/src/electoral_systems/block_witnesser.rs

```diff
@@ -172,123 +169,104 @@ impl<
-				);
-				CorruptStorageError::new()
-			}
-		);
-
-		// if the start of the range is
-
 		let BlockWitnesserState {
 			mut last_block_election_emitted_for,
-			last_block_received,
 			mut open_elections,
 			mut unprocessed_data,
 		} = ElectoralAccess::unsynchronised_state()?;
 
-		// e.g. the last block received is 20, we now have a range 19..21. This means, 19 and 20 are
-		// reorg'd.
-		log::info!("Last block received: {:?}", last_block_received);
-		log::info!("Witness range is: {:?}", witness_range);
-
-		// How do we handle the case where there's a reorg on the same block. We receive 4 and then
-		// we get 4 again.
-		if *witness_range.start() <= last_block_received {
-			log::info!("Reoooooorg");
-			// All ongoing elections are now invalid, we will recreate the elections, once the block
-			// height witnesser passes throught those block heights again, so the engines will
-			// revote.
-			election_identifiers.into_iter().for_each(|election_identifier| {
-				ElectoralAccess::election_mut(election_identifier).delete();
-			});
-
-			// Create a new election for each one in the range?
-
-			for root in
-				witness_range.clone().step_by(Into::<u64>::into(Chain::WITNESS_PERIOD) as usize)
-			{
-				log::info!("New election for root: {:?}", root);
-				ElectoralAccess::new_election(
-					(),
-					(
-						Chain::block_witness_range(root).into(),
-						ElectionGenerator::generate_election_properties(root),
-					),
-					(),
-				)?;
-				last_block_election_emitted_for = root;
-				open_elections = open_elections.saturating_add(1);
-			}
-
-			// NB: We do not clear any of the unprocessed data here. This is because we need to
-			// prevent double dispatches. By keeping the state, if we have a reorg we can check
-			// against the state in the process_block_data hook to ensure we don't double
-			// dispatch.
-		} else {
-			// ==== No reorg case ====
+		let mut remaining_election_identifiers = election_identifiers.clone();
 
-			// We could have multiple elections going, for different block/ranges.
-			for election_identifier in election_identifiers {
-				let election_access = ElectoralAccess::election_mut(election_identifier);
-				if let Some(block_data) = election_access.check_consensus()?.has_consensus() {
-					let (root_block_number, _extra_properties) = election_access.properties()?;
+		let last_seen_root = match chain_progress {
+			ChainProgress::WaitingForFirstConsensus => return Ok(()),
+			ChainProgress::Reorg(reorg_range) => {
+				// Delete any elections that are ongoing for any blocks in the reorg range.
+				for (i, election_identifier) in election_identifiers.into_iter().enumerate() {
+					let election = ElectoralAccess::election_mut(election_identifier);
+					let properties = election.properties()?;
+					if properties.0.into_range_inclusive() == *reorg_range {
+						election.delete();
+						open_elections = open_elections.saturating_sub(1);
+						remaining_election_identifiers.remove(i);
+					}
+				}
```

Comment:
Actually, when I detect a reorg, I have two ranges: `removed` and `added`. Depending on circumstances either one could be "longer" then the other. `removed` are the headers which I have previously witnessed and forwarded to you and are now invalid due to the reorg, `added` are the replacement headers now valid after the reorg.

It seems to me that since you want to delete the previously created elections, and recreate elections for the new blocks, you actually want to get both ranges. So I propose to change the variant to be `ChainProgress::Reorg { removed_range: InclusiveRange<>, added_range: InclusiveRange<> }`